snapcraft/linters/metadata_linter.py

```diff
@@ -0,0 +1,223 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+"""Snapcraft metadata linter.
+
+Validates that the snap's ``snapcraft.yaml`` (represented by ``SnapMetadata``)
+contains the minimum viable set of metadata fields and that their values look
+sane.  The checks are grouped into *ranks* so the user can prioritise the
+highest‑impact fixes first.
+
+Rank‑to‑severity mapping
+========================
+* **Rank‑1** – *Fatal* (build fails)
+* **Rank‑2** – *Error* (build continues but exits non‑zero)
+* **Rank‑3** – *Warning* (informational)
+"""
+
+import enum
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any, Optional
+
+from overrides import overrides
+
+from snapcraft.meta.snap_yaml import SnapMetadata
+
+from .base import Linter, LinterIssue, LinterResult
+
+_HELP_URL = "https://documentation.ubuntu.com/snapcraft/stable/reference/project-file/snapcraft-yaml/"
+
+
+class Rank(enum.IntEnum):
+    """Relative importance of a metadata field."""
+
+    RANK1 = 1
+    RANK2 = 2
+
+    @property
+    def severity(self) -> LinterResult:
+        if self is Rank.RANK1:
+            return LinterResult.WARNING
+        return LinterResult.INFO
+
+
+@dataclass
+class MetadataField:
+    name: str
+    rank: Rank
+    extract: Callable[["SnapMetadata"], Any]
+    lint: bool = True
+
+
+# Helper to pull the matching attribute values from all the apps
+def _get_apps_attr(meta: SnapMetadata, key: str) -> dict[str, list[str]] | None:
+    if not meta.apps:
+        return None
+
+    values: dict[str, list[str]] = {}
+    for name, app in meta.apps.items():
+        value: str | list[str] | None = getattr(app, key, None)
+        if value and isinstance(value, list):
+            values[name] = value
+        elif isinstance(value, str):
+            values[name] = [value]
+        else:
+            values[name] = []
+
+    return values if values else None
+
+
+# Helper to pull the matching attribute values from the links
+def _get_links_attr(meta: SnapMetadata, key: str) -> list[str] | str | None:
+    if not meta.links:
+        return None
+
+    return getattr(meta.links, key, None)
+
+
+_FIELDS: list[MetadataField] = [
+    # TODO: implement desktop field in Rank1
+    # TODO: implement icon field in Rank1
```

Comment:
I'm not sure these are good to have as warnings without further restrictions. The icon field makes sense to warn about for a desktop app, but for a shell program it may only make sense as an info level. For a kernel, it's not something that's even worth including an icon for.

Reply:
Icons are useful for any snap for the store listing, even if it is a Kernel snap, IMO. But there is no way of fetching the project icon.

> The icon field makes sense to warn about for a desktop app

This was also my initial concern about the desktop field. So, to solve it, I proposed an internal key, named `desktop`, which can be used as a checker, or maybe we can update snapd to support a lot more types.